r/R/dplyr-join.R

```diff
@@ -72,7 +72,18 @@ right_join.arrow_dplyr_query <- function(x,
                                          suffix = c(".x", ".y"),
                                          ...,
                                          keep = FALSE) {
-  do_join(x, y, by, copy, suffix, ..., keep = keep, join_type = "RIGHT_OUTER")
+
+  # Initially keep join keys so we can coalesce them after when keep=FALSE
+  query <- do_join(x, y, by, copy, suffix, ..., keep = TRUE, join_type = "RIGHT_OUTER")
+
+  # If we are doing a right outer join and not keeping the join keys of
+  # both sides, we need to coalesce. Otherwise, rows that exist in the
+  # RHS will have NAs for the join keys.
+  if (!keep) {
+    query$selected_columns <- post_join_projection(names(x), names(y), handle_join_by(by, x, y), suffix)
+  }
```

Comment:
This works, but it might be straightforward to fix here:

https://github.com/apache/arrow/blob/785ab5f3c3aa4e5f2e30655d4765ca729bee7aa2/r/R/dplyr-join.R#L37-L43

in the case of right join, we need to do the `setdiff` to produce the `left_output` and leave `right_output` as `names(y)`.

Reply:
I tried the way you suggested, but I think I still need the `post_join_projection()` call; if I revert what I've done and just make the changes suggested above, I get an error in my test as the schema expects `some_grouping.x` and `some_grouping.y` columns to have been created, but they have not.

Reply:
Try this (works for me locally):

```R
  left_output <- if (!keep && join_type == "RIGHT_OUTER") {
    setdiff(names(x), by)
  } else {
    names(x)
  }
  right_output <- if (keep || join_type %in% c("FULL_OUTER", "RIGHT_OUTER")) {
    names(y)
  } else {
    setdiff(names(y), by)
  }
```

Reply:
Thanks, I was missing the `!keep!` from the `left_output`!